### ISSUE TITLE
cytobank_to_gatingset only transform gated channels

### ISSUE THREAD

**Issue by bjreisman** [RGLab/CytoML#79]
Hi Mike + CytoML team, 

I've been working on re-scripting my workflow to work with the new API for gating and I've run into some undesirable behavior with regards to transformations. 

Basically, it appears that transformations are only applied to some of the channels in the gatingset, specifically those that are gated on. Meanwhile the other channels remain untransformed. In the past, my workflow was import-> compensate ->transform->gate, but if I try to transform the gatingset produced by `cytobank_to_gatingset`; some channels get transformed twice. 

I imagine this is kind of tricky to solve as cytobank allows the compensations/transformations use for visualization to be separated from the ones used for gating. Technically, the transformations in the gatingML file are the one's neccessary to reproduce the gating. Perhaps you could add additional arguments to `cytobank_to_gatingset` allowing the user to supply a transformerlist and compensation matrix or alternatively it may be preferable to just untransform the output of `cytobank_to_gatingset` and allow a custom set of transformations to be applied (as proposed here: https://github.com/RGLab/flowWorkspace/issues/301). 

I've tried to produce some reproducible code below:
```{r}
> library(CytoML)
> library(ggplot2)
> library(dplyr)
> library(flowCore)
> #this method of importing doesn't seem to work, but that's a seperate issue
> ce <- CytoML::open_cytobank_experiment("data2/experiment_34218_Feb-16-2020_08-33-PM.acs")
Unpacking ACS file...
> gs <- cytobank_to_gatingset(ce)
intact
singles
Live
All FCS files have the same following channels:
FSC-A
FSC-H
FSC-W
SSC-A
SSC-H
SSC-W
Alexa 647-A
Alexa Fluor 700-A
Time
write bachmabo_reismab_021420_Compensation Controls_Alexa 647 Stained Control_008.fcs to empty cdf slot...
write bachmabo_reismab_021420_Compensation Controls_Alexa Fluor 700 Stained Control_009.fcs to empty cdf slot...
write bachmabo_reismab_021420_Compensation Controls_Unstained Control_007.fcs to empty cdf slot...
write bachmabo_reismab_021420_Tonsil_0_10_005.fcs to empty cdf slot...
write bachmabo_reismab_021420_Tonsil_0_25_004.fcs to empty cdf slot...
write bachmabo_reismab_021420_Tonsil_0_50_003.fcs to empty cdf slot...
write bachmabo_reismab_021420_Tonsil_1_00_002.fcs to empty cdf slot...
write bachmabo_reismab_021420_Tonsil_2_00_001.fcs to empty cdf slot...
write bachmabo_reismab_021420_Tonsil_Unstained_006.fcs to empty cdf slot...
done!
.........done!
intact
singles
Live
.........done!
Error in list_to_dataframe(res, attr(.data, "split_labels"), .id, id_as_factor) : 
  Results do not have equal lengths
> 
> #
> fcs.paths <- list.files("data2", pattern = ".fcs", full.names = T)
> xml.path <- list.files("data2", pattern = ".xml", full.names = T)
> 
> gs <- CytoML::cytobank_to_gatingset(xml.path, fcs.paths)
intact
singles
Live
All FCS files have the same following channels:
FSC-A
FSC-H
FSC-W
SSC-A
SSC-H
SSC-W
Alexa 647-A
Alexa Fluor 700-A
Time
write bachmabo_reismab_021420_Tonsil_0_10_005.fcs to empty cdf slot...
write bachmabo_reismab_021420_Tonsil_0_25_004.fcs to empty cdf slot...
write bachmabo_reismab_021420_Tonsil_0_50_003.fcs to empty cdf slot...
write bachmabo_reismab_021420_Tonsil_1_00_002.fcs to empty cdf slot...
write bachmabo_reismab_021420_Tonsil_2_00_001.fcs to empty cdf slot...
write bachmabo_reismab_021420_Tonsil_Unstained_006.fcs to empty cdf slot...
done!
......done!
intact
singles
Live
......done!
> 
> orig_pdata <- pData(gs)
> orig_pdata$conc <- c(0.1, 0.25, 0.5, 1, 2, 0)
> pData(gs) <- orig_pdata
> 
> plotgs <- gs
> attr(plotgs, "subset") <- "singles"
> 
> 
> plotgs %>%
+   ggplot(aes(x=`Alexa Fluor 700-A`, y = `Alexa 647-A`)) +
+   geom_bin2d(bins = 128) +
+   scale_fill_viridis_c(option = "A") +
+   facet_wrap(conc~.) +
+   labs(title = "Ax700 transformed, Ax647 untransformed")
```
![image](https://user-images.githubusercontent.com/22528952/74612891-49831080-50cf-11ea-8fc9-0ac446df1f27.png)

```{r}
> 
> gs.transformed <-transform(gs, readRDS("data2/cytobank_transforms.rds"))
> 
> plotgs <- gs.transformed
> attr(plotgs, "subset") <- "singles"
> 
> plotgs %>%
+   ggplot(aes(x=`Alexa Fluor 700-A`, y = `Alexa 647-A`)) +
+   geom_bin2d(bins = 128) +
+   scale_fill_viridis_c(option = "A") +
+   facet_wrap(conc~.) +
+   labs(title = "Ax700 transformed twice, Ax647 transformed once")
```
![image](https://user-images.githubusercontent.com/22528952/74612896-5b64b380-50cf-11ea-9108-9e9b7ccc8b5e.png)

```{r}
> sessionInfo()
R version 3.6.2 (2019-12-12)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] flowCore_1.52.1 dplyr_0.8.4     ggplot2_3.2.1   CytoML_1.12.0  

loaded via a namespace (and not attached):
 [1] ncdfFlow_2.32.0      matrixStats_0.55.0   openCyto_1.24.0      RColorBrewer_1.1-2   httr_1.4.1          
 [6] Rgraphviz_2.30.0     tools_3.6.2          R6_2.4.1             KernSmooth_2.23-16   lazyeval_0.2.2      
[11] BiocGenerics_0.32.0  colorspace_1.4-1     flowWorkspace_3.34.1 withr_2.1.2          tidyselect_1.0.0    
[16] gridExtra_2.3        mnormt_1.5-6         curl_4.3             compiler_3.6.2       graph_1.64.0        
[21] Biobase_2.46.0       flowClust_3.24.0     ggcyto_1.14.0        labeling_0.3         flowStats_3.44.0    
[26] cytotidyr_0.0.1.100  scales_1.1.0         DEoptimR_1.0-8       hexbin_1.28.1        mvtnorm_1.0-12      
[31] robustbase_0.93-5    RBGL_1.62.1          stringr_1.4.0        digest_0.6.24        R.utils_2.9.2       
[36] rrcov_1.5-2          base64enc_0.1-3      jpeg_0.1-8.1         pkgconfig_2.0.3      htmltools_0.4.0     
[41] fastmap_1.0.1        rlang_0.4.4          rstudioapi_0.11      shiny_1.4.0          farver_2.0.3        
[46] jsonlite_1.6.1       mclust_5.4.5         gtools_3.8.1         R.oo_1.23.0          magrittr_1.5        
[51] Matrix_1.2-18        Rcpp_1.0.3           munsell_0.5.0        lifecycle_0.1.0      R.methodsS3_1.8.0   
[56] stringi_1.4.5        yaml_2.2.1           MASS_7.3-51.4        zlibbioc_1.32.0      CytobankAPI_1.3.0   
[61] plyr_1.8.5           grid_3.6.2           parallel_3.6.2       promises_1.1.0       crayon_1.3.4        
[66] lattice_0.20-38      splines_3.6.2        pillar_1.4.3         fda_2.4.8            corpcor_1.6.9       
[71] stats4_3.6.2         XML_3.99-0.3         glue_1.3.1           latticeExtra_0.6-29  data.table_1.12.8   
[76] RcppParallel_4.4.4   png_0.1-7            httpuv_1.5.2         gtable_0.3.0         purrr_0.3.3         
[81] clue_0.3-57          assertthat_0.2.1     ks_1.11.7            mime_0.9             janitor_1.2.1       
[86] xtable_1.8-4         later_1.0.0          viridisLite_0.3.0    IDPmisc_1.1.20       pcaPP_1.9-73        
[91] tibble_2.1.3         flowViz_1.50.0       ellipse_0.4.1        cluster_2.1.0  
```

vs. what I see in cytobank:
![image](https://user-images.githubusercontent.com/22528952/74612968-e2199080-50cf-11ea-9c70-3d44085c1c1d.png)

I've included the files necessary to run this example here: 
[experiment_34218.zip](https://github.com/RGLab/CytoML/files/4210939/experiment_34218.zip)

**Comment by mikejiang** [comment]
Right, the gatingML file only instructed `700-A` to be transformed but `647-A` was not. The initial goal of `cytobank_to_gatingset` (i.e. `cytobank_to_gatingset.default` or gatingML parser) was to replicate the gates and populations from cytobank. 
```r
gs <- cytobank_to_gatingset("~/Downloads/experiment_34218/CytExp_34218_Gates_v2.xml", files = "~/Downloads/experiment_34218/bachmabo_reismab_021420_Tonsil_0_10_005.fcs")
autoplot(gs, "Live") + scale_fill_viridis_c(option = "A")
```
![image](https://user-images.githubusercontent.com/1385649/74770786-ca621980-5241-11ea-8956-272559d982bf.png)

So in that sense, it is correct/or expected behavior to not touch the channels that aren't instructed by gatingML.
```r
autoplot(gs_cyto_data(gs), "700", "647")  + scale_fill_viridis_c(option = "A")
```
![image](https://user-images.githubusercontent.com/1385649/74770882-fed5d580-5241-11ea-8d66-86b50a93666a.png)
For visualizing the un-gated channels, you can easily use inline transformation provided by `ggcyto`
```r
cs <- gs_cyto_data(gs)
autoplot(cs, "700", "647")  + scale_fill_viridis_c(option = "A") + scale_y_flowjo_fasinh()
```
![image](https://user-images.githubusercontent.com/1385649/74770739-aef70e80-5241-11ea-99ae-8445c3ceea81.png)
Or manually transform the `gs` , make sure only transform the specified channels so that the other channels won't be transformed twice
```r
gs <- transform(gs, transformerList("Alexa 647-A", asinhtGml2_trans()))
cs <- gs_cyto_data(gs)
autoplot(cs, "700", "647")  + scale_fill_viridis_c(option = "A")
```
![image](https://user-images.githubusercontent.com/1385649/74771612-4d37a400-5243-11ea-9587-28342da263cd.png)

 I do see the `scales` section in yaml file from cytobankExperiment file (acs). Potentially we could parse `transformations` from yaml (i.e. in `cytobank_to_gatingset.cytobank_experiment`) and overwrite/ignore the ones from gatingML file. But we don't know if there will be potential discrepancy between the two thus cause the trouble for gating itself.  So if the solutions I provided above are sufficient to address your viz problem, then it may not be worth to go that direction at this point.

That said, I am going to look into the acs parsing error.

**Comment by mikejiang** [comment]
>Error in list_to_dataframe(res, attr(.data, "split_labels"), .id, id_as_factor) : 
  Results do not have equal lengths

It was due to the missing `tags` (i.e. `doses`) for some files. I've pushed the quick fix to fill these missing values with NA.

**Comment by mikejiang** [comment]
I attempt to implement the idea of using `scales` from yaml instead of gatingML (in `scale_from_yaml` branch) 
Here is the transformation in gatingML
```xml
    <transforms:fasinh transforms:T="1175.2011936438014" transforms:M="0.43429448190325176" transforms:A="0.0" />
```
In order to replicate it, we need these 3 parameters or their equivalents
And here is the scale from yaml
```yaml
- !org.cytobank.pipeline.experiment.Scale
   channelShortName: alexa fluor 700-a
   cofactor: 1000
   maximum: 262144.0
   minimum: -2000.0
   scaleType: Arcsinh
```
`T` can be computed from `cofactor` : `sinh(1) * 1000 = 1175.201`
But it is not clear to me how to compute  `M` value (i.e. numeric the full width of the transformed display in asymptotic decades) from the rest of given numbers.

**Comment by mikejiang** [comment]
I have a hypothesis that `M=0.43429448190325176` and `A=0` are constant values for cytobank given  there are no equivalent found in yaml version.  It seems to parse the gates ok (from my limited test cases). 
Since you are the primary user of ACS parser so far, I will go ahead to merge this feature to `master` and let you decide whether it works for you. Here is the result which has the ungated channel transformed as well
```r
ce_get_transformations(ce)
gs <- cytobank_to_gatingset(ce)
gs <- subset(gs, Doses != "NA")
cs <- gs_cyto_data(gs, "singles")
ggcyto(cs, aes("700", "647")) + geom_bin2d(bins = 128)  + scale_fill_viridis_c(option = "A") + facet_wrap(Doses~.)
```
![image](https://user-images.githubusercontent.com/1385649/74786803-76b3f800-5262-11ea-8156-8d00be6fa08e.png)

Note that I don't plan to push this feature to bioc release (since it is not a bug fix), so you will need to upgrade all your flow tools from bioc devel or github

**Comment by SamGG** [comment]
Well done!
Note sure it helps:
`log10(exp(1)) - 0.43429448190325176 = 5.55e-17`

**Comment by bjreisman** [comment]
Awesome! This looks like a very useful feature (for my workflow). Cytobank is odd in that it allows you manipulate scales and transformations independently from gates such that you create figures on a different set of scales/transformations that your gates were drawn, but the scales/transformations used at the time of gating stay tied to the gates. That's my understanding at least. I think your solution is optimal as it handles gates/transformations the same way cytobank does. I haven't yet tried to import an experiment where there are two sets of gates drawn on different transformations...

I noticed the asinh transformation in flowworkspace has a few more parameters than the one [used in cytobank](https://support.cytobank.org/hc/en-us/articles/206148057-About-the-Arcsinh-transform).  If you're interested I wrote a transformation function that the cytobank asinh transformation [here](https://github.com/bjreisman/cytotidyr/blob/master/R/asinh_trans_prettybreaks.R), though your solution appears to get to the same results (at least to 10e-17 😄). 

I've had some troubling getting flowworkspace installed on my bioc-devel/R4.0 environment, I think due to this [issue](https://github.com/RGLab/cytolib/issues/31), but I'll make another attempt this weekend and try this out. 

Thanks again!
